docs/dependencies.md

```diff
@@ -0,0 +1,287 @@
+two major places:
+* For production, in our puppet configuration, `puppet/zulip/`, using
+  the `Package` and `SafePackage` directives.  For development, in
+  `APT_DEPENDENCIES` in `tools/lib/provision.py`.
+* The packages needed to build a Zulip virtualenv, in
+  `VENV_DEPENDENCIES`.  These are separate from the rest because we
+  may need to install a virtualenv before running the more complex
+  scripts that, in turn, install other dependencies, and also because
+  that list is shared between development and production.
+
+We maintain a [personal package archive (PPA)][ppa] with some packages
+unique to Zulip (e.g the `tsearch_extras` postgres extension) and
+backported versions of other dependencies (e.g. `camo`, to fix a buggy
+`init` script).  Our goal is to shrink or eliminate this PPA where
+possible by getting issues addressed in the upstream distributions.
+
+We also rely on the `pgroonga` PPA for the `pgroonga` postgres
+extension, used by our [full-text search](full-text-search.html).
+
+## Python packages
+
+We manage Python packages via the Python-standard `requirements.txt`
+system and virtualenvs, but there’s a number of interesting details
+about how Zulip makes this system work well for us that are worth
+highlighting.  The system is largely managed by the code in
+`scripts/lib/setup_venv.py`
+
+* **Using `pip` to manage dependencies**.  This is standard in the
+  Python ecosystem, and means we only need to record a list of
+  versions in a `requirements.txt` file to declare what we're using.
+  Since we have a few different installation targets, we maintain
+  several `requirements.txt` format files in the `requirements/`
+  directory (e.g. `dev.txt` for development, `prod.txt` for
+  production, `docs.txt` for ReadTheDocs, `common.txt` for the vast
+  major of packages common to prod and development, etc.).  We use
+  `pip install --no-deps` to ensure we only install the packages we
+  explicitly declare as dependencies.
+* **virtualenv with pinned versions**.  For a large application like
+  Zulip, it is important to ensure that we're always using consistent,
+  predictable versions of all of our Python dependencies.  To ensure
+  this, we install our dependencies in a [virtualenv][] that contains
+  only the packages and versions that Zulip needs, and we always pin
+  exact versions of our dependencies in our `requirements.txt` files.
+  We pin exact versions, not minimum versions, so that installing
+  Zulip won't break if a dependency makes a buggy release.  A side
+  effect is that it's easy to debug problems caused by dependency
+  upgrades, since we're always doing those upgrades with an explicit
+  commit updating the `requirements/` directory.
+* **Caching of virtualenvs and packages**.  To make updating the
+  dependencies of a Zulip installation efficient, we maintain a cache
+  of virtualenvs named by the hash of the relevant `requirements.txt`
+  file.  That way, when re-provisioning a development environment or
```

Comment:
... and the requirements files it includes, e.g. `common.txt`

Reply:
Oh hmm but maybe that part is effectively dead code now, thanks to the "lock" files?

I think it'd be helpful to move the "Pinning versions of indirect dependencies" bullet above this one -- it puts the discussion of pinning together, and it sets up the right picture for what's happening by the time we're installing from these dependencies, which is what this and the next bullet are about.

Reply:
Also the `--no-deps` mention above is confusing before the lock files have been explained -- we don't manually write out all our indirect dependencies, because that'd be a pain, but we do pin them. I think I'd move that sentence down to the graf about pinning indirect dependencies.